sdk/core/azure-core-experimental/azure/core/experimental/transport/_urllib3.py

```diff
@@ -0,0 +1,300 @@
+import os
+import functools
+from typing import Any, Callable, Mapping, Optional, Union, Dict, cast, List, Tuple, Iterator
+
+import urllib3
+
+from azure.core.pipeline import Pipeline
+from azure.core.exceptions import ServiceRequestError, ServiceResponseError, IncompleteReadError, DecodeError
+from azure.core.configuration import ConnectionConfiguration
+from azure.core.pipeline.transport import HttpTransport
+from azure.core.rest import HttpRequest as RestHttpRequest, HttpResponse as RestHttpResponse
+from azure.core.rest._http_response_impl import HttpResponseImpl
+
+
+DEFAULT_BLOCK_SIZE = 32768
+_TYPE_FIELD_VALUE = Union[str, bytes]
+_TYPE_FIELD_VALUE_TUPLE = Union[
+    _TYPE_FIELD_VALUE,
+    Tuple[str, _TYPE_FIELD_VALUE],
+    Tuple[str, _TYPE_FIELD_VALUE, str],
+]
+
+
+def _read_files(fields, files):
+    if files is None:
+        return
+    if isinstance(files, Mapping):
+        files = list(files.items())
+    for k, v in files:
+        if hasattr(v, "read"):
+            updated = v.read()
+        elif isinstance(v, tuple):
+            updated = tuple(i.read() if hasattr(i, "read") else i for i in v)
+        else:
+            updated = v
+        fields.append((k, updated))
+
+
+class Urllib3StreamDownloadGenerator:
+    """Generator for streaming response data.
+
+    :param pipeline: The pipeline object
+    :param response: The response object.
+    :keyword bool decompress: If True which is default, will attempt to decode the body based
+        on the *content-encoding* header.
+    """
+
+    def __init__(self, pipeline: Pipeline, response: "Urllib3TransportResponse", **kwargs: Any) -> None:
+        self.pipeline = pipeline
+        self.response = response
+        decompress: bool = kwargs.pop("decompress", True)
+        self.iter_content_func: Iterator[bytes] = self.response._internal_response.stream(
+            amt=self.response._block_size, decode_content=decompress
+        )
+
+    def __iter__(self) -> "Urllib3StreamDownloadGenerator":
+        return self
+
+    def __next__(self) -> bytes:
+        try:
+            return next(self.iter_content_func)
+        except urllib3.exceptions.DecodeError as err:
+            raise DecodeError(err, error=err) from err
+        except (urllib3.exceptions.IncompleteRead, urllib3.exceptions.InvalidChunkLength) as err:
+            raise IncompleteReadError(err, error=err) from err
+        except urllib3.exceptions.HTTPError as err:
+            msg = err.__str__()
+            if "IncompleteRead" in msg:
+                raise IncompleteReadError(err, error=err) from err
+            raise ServiceResponseError(err, error=err) from err
+
+
+class Urllib3TransportResponse(HttpResponseImpl):
+    def __init__(
+        self, *, request: RestHttpRequest, internal_response: urllib3.response.HTTPResponse, **kwargs: Any
+    ) -> None:
+        headers = kwargs.pop("headers", internal_response.headers)
```

Comment:
The response headers here would probably need to be wrapped in some kind of compatibility layer if we need them to be 100% backwards compatible with `requests`.

Reply:
I don't think we want to be 100% backward compatible with `requests`. We own the definition of what headers are (e.g. multi-key-case-insensitive that normalizes lists of values/multiple instances of keys)

Reply:
This has been resolved by customizing the `ItemsView` of the returned headers to combine multi-headers rather than keep them separate.